cloud_storage/google_cloud_storage/download_file.py

```diff
@@ -0,0 +1,220 @@
+
+    blob.download_to_filename(local_path)
+
+    print(f'{blob.bucket.name}/{blob.name} successfully downloaded to {local_path}')
+
+    return
+
+
+def get_gclient():
+    """
+    Attempts to create the Google Cloud Storage Client with the associated
+    environment variables
+    """
+    try:
+        gclient = storage.Client()
+    except Exception as e:
+        print(f'Error accessing Google Cloud Storage with service account ' \
+                f'{args.gcp_application_credentials}')
+        raise(e)
+
+    return gclient
+
+
+def get_bucket(*,
+        gclient,
+        bucket_name):
+    """
+    Fetches and returns the bucket from Google Cloud Storage
+    """
+    try:
+        bucket = gclient.get_bucket(bucket_name)
+    except NotFound as e:
+        print(f'Bucket {bucket_name} does not exist\n {e}')
+        raise(e)
+
+    return bucket
+
+
+def main():
+    args = get_args()
+    set_environment_variables(args)
+    bucket_name = args.bucket_name
+    source_file_name = args.source_file_name
+    source_folder_name = clean_folder_name(args.source_folder_name)
+    source_full_path = combine_folder_and_file_name(
+        folder_name=source_folder_name, file_name=source_file_name)
+    source_file_name_match_type = args.source_file_name_match_type
+
+    destination_folder_name = clean_folder_name(args.destination_folder_name)
+    if not os.path.exists(destination_folder_name) and \
+            (destination_folder_name != ''):
+        os.makedirs(destination_folder_name)
+
+    gclient = get_gclient()
+    bucket = get_bucket(gclient=gclient, bucket_name=bucket_name)
+
+    if source_file_name_match_type == 'regex_match':
+        file_names = find_google_cloud_storage_file_names(bucket=bucket,
+                                            prefix=source_folder_name)
+        print(file_names)
+        matching_file_names = find_matching_files(file_names,
+                                            re.compile(source_file_name))
+        print(f'{len(matching_file_names)} files found. Preparing to download...')
+
+        for index, blob in enumerate(matching_file_names):
+            destination_name = determine_destination_name(
+                    destination_folder_name=destination_folder_name,
+                    destination_file_name=args.destination_file_name,
+                    source_full_path=blob.name, file_number=index+1)
+
+            print(f'Downloading file {index+1} of {len(matching_file_names)}')
+            download_google_cloud_storage_file(blob=blob,
+                    destination_file_name=destination_name)
+    else:
+        destination_name = determine_destination_name(
+                destination_folder_name=destination_folder_name,
+                destination_file_name=args.destination_file_name,
+                source_full_path=source_full_path)
+
+        download_google_cloud_storage_file(blob=blob,
+                destination_file_name=destination_name)
```

Comment:
This is returning `UnboundLocalError: local variable 'blob' referenced before assignment` when you run with `exact_match` under the argument `source-file-name-match-type`

Reply:
Fixed. Looks like when I was testing I only tested the regex matching downloads..oops

Also added support for json string input and updated the GCS python module!